woocommerce/woocommerce-gutenberg-products-block
# Track checkout and cart page info (blocks in use or shortcode in use)

## What

This PR adds tracking for cart and checkout pageview events when using Jetpack WooCommerce Analytics.

Summary of changes:

- Add new JS script, `wc-blocks-jetpack-woocommerce-analytics` which is only loaded when on Cart/Checkout or order received pages.
- Add `JetpackWooCommerceAnalytics` class to handle loading of the above script when on the cart/checkout page.
- Add checks to sense whether cart/checkout shortcodes or blocks are in use. Still necessary because even though we're in WC Blocks, a merchant may be using the shortcode on one page and the block on the other.
- Add functions that dispatch the page view events in JavaScript.

### Why we are using `DOMContentLoaded`

We added an event listener for `DOMContentLoaded` in the above mentioned `wc-blocks-jetpack-woocommerce-analytics` script that would run the `maybeTrack...PageView` functions. While relying on `DOMContentLoaded` doesn't feel great, it's probably the only way we can reliably send the tracking events when `_wca.push` is ready, below is a list of issues we faced trying to get this working:
  - When tracking cart page view event, we tried to subscribe to the `wc/store/cart` data store, and send the tracking event as soon as the cart was loaded. Unfortunately, it seemed that the cart data store finished initialising and had received all of its actions before the `_wca.push` function was ready.
  - When tracking the checkout page view event, we _could_ rely on a subscription to the `wc/store/cart` data store to send the event. This is because the store would receive an action (`SELECT_SHIPPING_RATE`) after page load which caused our subscribe function to run and send the event, however if the cart did not have any shippable items this would not have fired.
  - This method of subscribing to the data store is not possible on the order received page as no data store is loaded there.

### Order Received event

This event is a little different to the cart/checkout pageview events because at the point of recording those events we have access to the `wc/store/cart` data store and therefore we are able to get the cart items easily to fill the event properties. In the order received page, we don't have this data store, and no way to access the data on the front end. 

The approach taken here was to collect all of the data relating to an order on the `woocommerce_thankyou` hook in PHP, and add it to the WooCommerce Blocks asset data registry. Since this hook only fires on the order received page, that key in the asset data registry would only be set on that page. On the front end we can check for the presence of data in that key and therefore determine reliably if we are on the order received page.

The page would load and this key that contains the event properties would be used to populate a `_wca.push` call.

## Why

To enable pageview tracking when using the WooCommerce Analytics module in Jetpack.



## Testing Instructions



_Please consider any edge cases this change may have, and also other areas of the product this may impact._

0. Install Tracks Vigilante p7H4VZ-3cB-p2
1. Enable Jetpack and connect it WPCOM.
2. Add the Cart block to the cart page, add the checkout shortcode to the checkout page.
3. Add an item to your cart and go to the Cart page.
4. Ensure a pageview event is sent for the cart in tracks vigilante. 
5. Ensure the properties: `cart_page_contains_cart_block`, `cart_page_contains_cart_shortcode`, `checkout_page_contains_checkout_block`, and `checkout_page_contains_checkout_shortcode` are correct.
6. Change the checkout page to use the Checkout block.
7. Repeat step 4 and 5.
8. Visit the Checkout page, repeat 4 and 5 but expect the checkout page view event.
9. Add additional blocks to the cart/checkout pages and check they are reported properly in the event properties in tracks vigilante.
10. Go back to the Cart block and change the quantity of items, remove items, add coupons etc. Ensure no additional cart page view events are tracked.
11. On the order received page, verify that the `woocommerceanalytics_order_confirmation_view` event is being sent and the correct properties are sent with it.


## Screenshots or screencast



| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility


Required:


## Checklist

Required:

Conditional:

## Changelog


> Skipping
